### ISSUE TITLE
Looking for someone to help creating a sample addon using the Testimonials element

### ISSUE THREAD

**Issue by iamcanadian1973** [Theme-Fusion/Fusion-Builder-Sample-Add-On#44]
I've managed to get a stand alone working version, except I think I'm missing some Underscore templates and I'm not sure how to load them. Everything works except the Live frontend editing in the slide out panel. If I edit any of the values, it just disappears.

I've taken as much as I could from referencing the Hello world Sample addon, but I'm stuck.

I'd prefer not to post on Upwork or the like, as it never works out. So I was hoping to find someone here. I'd pay for your time. 

Once it's working I'd like to make the code public as well as I've searched everywhere and found it impossible to find an up to date addon example that is more complex. 

Thanks,

Kyle

**Comment by mikka23** [comment]
Hi Kyle, 

What do you have so far?  Also what is the target of your project.  With some specific details I can offer some advice/check what is going wrong.

**Comment by iamcanadian1973** [comment]
Mark,

Thanks so much. 

This is what I have so far. 

[fusion-builder-icertis-testimonials.zip](https://github.com/Theme-Fusion/Fusion-Builder-Sample-Add-On/files/4440819/fusion-builder-icertis-testimonials.zip)

Target of the project is building a testimonials slider with a custom layout on the front end for a client website. Sadly the clients needs changed, and I'll be switching to creating a stand alone addon based on a CPT possible something similar to a posts slider, but I have already spent 12 hours trying to get what i have working, so I figure from there I'd be able to understand many of the complex elements. 

I've copied over all the files I could figure out from the Fusion Builder element, and everything is working except the live refresh. I think I'm missing some underscore templates that I'm not sure how to load. 

Goal is to be able to create custom elements as needed when the existing elements are almost what is needed, but just need some additional fields or a custom layout. 

Thanks,

Kyle

**Comment by iamcanadian1973** [comment]
Michael,

I figured out how to get my two issues working (9 hours later :)). I've attached the updated plugin. 

[fusion-builder-icertis-testimonials.zip](https://github.com/Theme-Fusion/Fusion-Builder-Sample-Add-On/files/4442347/fusion-builder-icertis-testimonials.zip)

I did some review of fusion builder and figured out I was missing two items:

```
add_action( 'fusion_builder_enqueue_live_scripts', function() {
    wp_enqueue_script( 'fusion_builder_icertis_testimonial_element', ICERTIS_TESTIMONIALS_URL . 'elements/front-end/view.js', [], FUSION_BUILDER_VERSION, true );   
});

add_action( 'fusion_builder_load_templates', function() {
    include ICERTIS_TESTIMONIALS_DIR . 'elements/front-end/frontend.php';
});
```

My only remaining issues are:

1. Live editor preview Avatar does not show, and a console error
2. Frontend page where slider is working, no image is showing 

Would you be able to help figure out why those out? I really appreciate any help you can give me.

Thanks,

Kyle

**Comment by iamcanadian1973** [comment]
Michael,

I removed my JS error with:

```
// Missing template_category
add_action( 'fusion_builder_admin_scripts_hook', function () {
	wp_localize_script(
        'fusion_builder',
        'fusionBuilderConfig',
        [
            'template_category' => false,
        ]
    );
}, 15 );
```

**Comment by mikka23** [comment]
[fusion-builder-icertis-testimonials.zip](https://github.com/Theme-Fusion/Fusion-Builder-Sample-Add-On/files/4443470/fusion-builder-icertis-testimonials.zip)

Hi Kyle,

I made some alterations to your code and also added some comments starting with TIP (zip above).  So if you search that you will find them.  Mostly in the PHP, one in the JS view.  You could either continue with those or just check the comments and pick the parts you want.

In general what you have now is a little odd, but it really depends on what you want to achieve as to what I would recommend.  If its just the testimonial slider then this can work. You can edit from here to adjust to your liking and maybe add some options.  Depending on what you want you can continue to rely on fusion-testimonials.js or you could create your own version of that (more on that later).  If you are wanting to learn about the builder and build your own elements, then testimonials is probably not the best starting point.  Mainly because the PHP code there is a bit hard to follow and dated, and in general the parent/child ones are more complex.  So if you wanted something completely new then I would recommend using the front-end branch on this repo as a starting point instead and then building from there.

Regarding the specific problems you mentioned.  The main issues were:
1.  I wasn't actually replicating this.  So long as color was set the avatar was showing.  If it disappeared on edit, this is due to the JS trigger.  So when you make an edit, it re-renders the code then triggers the function afterPatch.  In your case that was calling - this._refreshJs().

What that function does is take the shortcode name and trigger a JS event which you can listen for in your regular front-end JS.  So in this example usually the testimonials element triggers fusion-element-render-**fusion_testimonials**.   This then triggers JS to turn the markup into a slider.  The same code which is used on the front-end.  You can find this in fusion-builder/assets/js/general/fusion-testimonials.js.

In your case I don't see any JS for the slider aspect, so I assume you meant to use the same as the fusion-testimonials.  In which case the trigger is wrong because it will instead trigger fusion-element-render-**icertis_testimonials**.  To resolve can just trigger correct one directly:
```js
jQuery( '#fb-preview' )[ 0 ].contentWindow.jQuery( 'body' ).trigger( 'fusion-element-render-fusion_testimonials', this.model.attributes.cid );
```

2.  The logic in the PHP was a bit weird and canceling out the image markup if an image was selected instead of avatar.

**Comment by iamcanadian1973** [comment]
Michael,

Thanks so much. I'm going to go through all your suggestions.

I won't be using this element in production, I'll just be learning from it. 

I chose the Testimonials because I wanted to understand the parent child elements and be able to build those. I had no clue how to even begin to understand that from the Sample Addon, so that's why I decided to go the direction I did. 

I definitely don't want to rely on the existing testimonials JS, so I'll lern how to adjust that as well.

I absolutely want to learn how to build custom elements correctly, for me it's just much easier to learn by deconstructing. 

With that said, I plan to learn from all your feedback and then take that knowledge into the Sample Addon and get it done correctly. 

I'll keep you posted and share what I come up with in the future.

Thanks again for all your help.

Where I struggle the most is the JS. So with the Testimonials example being old, are you saying that there is a better way and I should be replacing these two below?:

```
this.generateChildElements();

this._refreshJs();
```
Is there a better way I should be regenerating and refreshing the updates?

Should these two:

```
FusionPageBuilder.ParentElementView.extend
FusionPageBuilder.ChildElementView.extend
```

Now just be a single

```
FusionPageBuilder.ElementView.extend
```

Learning the JS is going to be the challenging part.

I realize that this below is just a hack?

```
add_action( 'fusion_builder_enqueue_live_scripts', function() {
    wp_enqueue_script( 'fusion_builder_icertis_testimonial_element', ICERTIS_TESTIMONIALS_URL . 'elements/front-end/view.js', [], FUSION_BUILDER_VERSION, true );
});


add_action( 'fusion_builder_load_templates', function() {
    include ICERTIS_TESTIMONIALS_DIR . 'elements/front-end/frontend.php';
});
```

But when I tried to add them using the below the refresh didn't work, and it gives me a console error that I can't see because of the frames. Is there a tip you have for accessing the JS errors from the frame? It just shows there is an error but you can't get to it (seeing these errors would be so helpful). 

See screenshot https://www.evernote.com/l/AOG30ULB6TlDf4RouE3hpVEqSF4Fe4jZ4Ok

```
front_end_custom_settings_view_js
front-end

```

Kyle

PS You're very generous with your help, is there a charity or organization I could donate to in your name to show my appreciation?

**Comment by mikka23** [comment]
Hi Kyle,

```js
this.generateChildElements();
```

First one you will want if its parent/child.  It's in the parent so that when an option changes all the children are then updated to use that new value.  So should be kept.

```js
this._refreshJs();
```

This depends.  The testimonial element is a slider, so it takes the HTML markup and then JS transforms that into a slider.  If the element is not dynamic like that, then you wouldn't need to call this at all.  Best way to think about is on a regular page load, what is required to get the element working?  For testimonial on a regular page load it usually turns into a slider via JS.  For an image carousel it becomes a scrolling carousel. If the element needs more JS to work, then this in the place to do it.  The trigger approach that _refreshJs does it helpful because often you have the other JS for page load already.  So it saves some time.

If your element does not need extra dynamic JS, for example if your testimonials were just boxes with text and images, then you actually don't need anything else here at all.

### View Naming
```js
FusionPageBuilder.ParentElementView.extend
FusionPageBuilder.ChildElementView.extend
```

These are still correct for parent and child elements.  The other one you referenced is for regular elements which are not parent/child.

### Loading View and Template
For loading the view and the template, what you have will work and is perfectly valid.  But defining it in the element map should also work and is a bit easier to follow (front-end branch of this repo uses that). For example: https://github.com/Theme-Fusion/Fusion-Builder-Sample-Add-On/blob/front-end/elements/hello-world.php#L204-L208.  That should work.  If it doesn't do you have an example?  Alternatively you can enqueue from the element class.

### Error Reporting
You should be able to see errors on both frames if it's a JS error.  Some are also intercepted and captured.  If that happens you should get a dialog with a button to copy the full details.  That essentially copies the backtrace for the error.  You might need to change frame in the developer tools from the area you have highlighted. (note, errors in underscore templates are harder to interpret).

### Testimonial Element
The main reason I mentioned not recommending the testimonial element is a few reason, but does not mean what the testimonial element is invalid:
1.  It requires JS for the element itself.  ie the slider part, that adds a bit more complication when testing things and means re-render is slower and not as easy to work with.
2.  The PHP is coded in a way that we usually would not recommend now (sprintf and various other things).  So overall it is just a bit harder to follow the code than I would like for a new element.
3. Parent/child elements are a lot more complicated for the underscore template + view.  Learning with a regular element is definitely easier and gives a clearer indication about how things work.

### More Information
There are some more details written on the main page here - https://github.com/Theme-Fusion/Fusion-Builder-Sample-Add-On/tree/front-end.  That was written rather quickly a while ago, but should give some more details about certain things which are not included in the documentation (which needs re-done for this.)



I work for ThemeFusion, so definitely no donation required 😀

**Comment by iamcanadian1973** [comment]
Michael,

Thanks so much. 

Just wanted to make sure nothing I was was at risk for being deprecated. 

Theme Fusion is lucky to have you!

Kyle

**Comment by iamcanadian1973** [comment]
Michael,

I'm trying to pull a list of terms in to the shortcode


```
function fusion_element_icertis_quotes() {
    
    $builder_status = function_exists( 'is_fusion_editor' ) && is_fusion_editor();
    
    var_dump( get_terms( array(
    'taxonomy' => 'quotes_type',
) ) );
    
	fusion_builder_map(
		fusion_builder_frontend_data(
			'Icertis_Quotes',
			[
				'name'          => esc_attr__( 'Icertis Quotes', 'fusion-builder' ),
				'shortcode'     => 'icertis_quotes',
				'icon'          => 'fusiona-bubbles',
                'front-end'     => ICERTIS_ELEMENTS_PLUGIN_DIR . 'elements/front-end/icertis-quotes.php',
                'front_end_custom_settings_view_js' => ICERTIS_ELEMENTS_PLUGIN_DIR . 'elements/front-end/icertis-quotes.js',
				'child_ui'      => false,
				'sortable'      => false,
				//'help_url'      => 'https://theme-fusion.com/documentation/fusion-builder/elements/quotes-element/',
				'params'        => [
                    [
						'type'        => 'multiple_select',
						'heading'     => esc_attr__( 'Categories', 'fusion-builder' ),
						'placeholder' => esc_attr__( 'Categories', 'fusion-builder' ),
						'description' => esc_attr__( 'Select categories of posts to display or leave blank for all.', 'fusion-builder' ),
						'param_name'  => 'cat_slug',
						'value'       => $builder_status ? fusion_builder_shortcodes_categories( 'quotes_type' ) : [],
						'default'     => '',
						/*'callback'    => [
							'function' => 'fusion_post_slider_query',
							'ajax'     => true,
						],*/
					],
					[
						'type'        => 'textfield',
						'heading'     => esc_attr__( 'CSS Class', 'fusion-builder' ),
						'description' => esc_attr__( 'Add a class to the wrapping HTML element.', 'fusion-builder' ),
						'param_name'  => 'class',
						'value'       => '',
						'group'       => esc_attr__( 'General', 'fusion-builder' ),
					],
					[
						'type'        => 'textfield',
						'heading'     => esc_attr__( 'CSS ID', 'fusion-builder' ),
						'description' => esc_attr__( 'Add an ID to the wrapping HTML element.', 'fusion-builder' ),
						'param_name'  => 'id',
						'value'       => '',
						'group'       => esc_attr__( 'General', 'fusion-builder' ),
					],
				],
			]
		)
	);
}
```

I had to change the hook to get it to work though. The Taxonomy is already registered in the Child theme functions.php file so I can't move it. 

Is there a better hook to use than below? 

```
//add_action( 'fusion_builder_before_init', 'fusion_element_icertis_quotes' );
add_action( 'wp_loaded', 'fusion_element_icertis_quotes', 99 );
```

**Comment by iamcanadian1973** [comment]
Michael,

how would I customize the flexslider variables?

The below doesn't work.

```
public function add_scripts() {
                Fusion_Dynamic_JS::enqueue_script( 'fusion-flexslider' );
                
                Fusion_Dynamic_JS::localize_script(
                    'fusion-flexslider',
                    'fusionFlexSliderVars',
                    [
                        'animation'              => 'slide',
                        'slideshow_autoplay'     => false,
                        'slideshow_speed'        => 5000,
                        //'flex_smoothHeight'      => $flex_smooth_height
                    ]
                );
			}
```

**Comment by mikka23** [comment]
Hi Kyle,

Changing the hook is fine.  You will notice for our portfolio element for example, we also use **wp_loaded**.

I wouldn't recommend changing the variables to the fusion-flexslider file.  Instead I would recommend just creating your own JS file.  You can still use the flexslider library, but the fusion-flexslider file is just for initiating various flexsiders in the theme.  The actual jQuery Flexslider is library is a different file (slug jquery-flexslider, filename jquery.flexslider.js).

So instead it would just be something like:

```php

Fusion_Dynamic_JS::enqueue_script(
    'icertis-flexslider',
    MY_PLUGIN_URL . '/js/incertis-flexslider.js',
    MY_PLUGIN_PATH . '/js/incertis-flexslider.js',
    [ 'jquery', 'jquery-flexslider' ],
    '1',
    true
);

 Fusion_Dynamic_JS::localize_script(
    'icertis-flexslider',
    'icertisFlexSliderVars',
    [
        'animation'              => 'slide',
        'slideshow_autoplay'     => false,
        'slideshow_speed'        => 5000,
    ]
);
```

Then you can put whatever you need in that JS file for initiating a slider and can pass whatever variables you want.  For example something like (pure example and not tested):
```js
// Init slider on page load (front-end.)
jQuery( window ).on( 'load', function() {
	jQuery( '.icertis-testimonials' ).flexslider( {
		slideshow: Boolean( Number( icertisFlexSliderVars.slideshow_autoplay ) ),
		slideshowSpeed: icertisFlexSliderVars.slideshow_speed,
		start: function( slider ) {
		}
	} );
} );

// Event fired when live editor reders the element, init slider for that specific area.
jQuery( window ).on( 'fusion-element-render-icertis_testimonials', function( $, cid ) {
	var $targetEl = 'undefined' !== typeof cid ? jQuery( 'div[data-cid="' + cid + '"]' ) : jQuery( document ),
		$slider   = $targetEl.find( '.icertis-testimonials' );

	if ( $slider && $slider.length ) {
		$slider.( '.icertis-testimonials' ).flexslider( {
			slideshow: Boolean( Number( icertisFlexSliderVars.slideshow_autoplay ) ),
			slideshowSpeed: icertisFlexSliderVars.slideshow_speed,
			start: function( slider ) {
			}
		} );
	}
} );
```
